crates/kdc/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+mod as_exchange;
+pub mod config;
+mod error;
+mod tgs_exchange;
+mod ticket;
+
+use std::time::Duration;
+
+use picky_asn1::wrapper::{ExplicitContextTag0, GeneralizedTimeAsn1, OctetStringAsn1};
+use picky_asn1_der::Asn1DerError;
+use picky_krb::constants::types::{NT_ENTERPRISE, NT_PRINCIPAL, NT_SRV_INST};
+use picky_krb::data_types::PrincipalName;
+use picky_krb::messages::{AsReq, KdcProxyMessage, TgsReq};
+use time::OffsetDateTime;
+
+use crate::as_exchange::handle_as_req;
+use crate::config::{DomainUser, KerberosServer};
+use crate::error::KdcError;
+use crate::tgs_exchange::handle_tgs_req;
+
+fn find_user_credentials<'a>(
+    cname: &PrincipalName,
+    realm: &str,
+    kdc_config: &'a KerberosServer,
+) -> Result<&'a DomainUser, KdcError> {
+    let username = if cname.name_type.0.0 == [NT_PRINCIPAL] {
+        let cname = &cname
+            .name_string
+            .0
+            .first()
+            .ok_or(KdcError::ClientPrincipalUnknown(
+                "the incoming KDC request does not contain client principal name".to_owned(),
+            ))?
+            .0;
+        format!("{cname}@{realm}")
+    } else if cname.name_type.0.0 == [NT_ENTERPRISE] {
+        cname
+            .name_string
+            .0
+            .first()
+            .ok_or(KdcError::ClientPrincipalUnknown(
+                "the incoming KDC request does not contain client principal name".to_owned(),
+            ))?
+            .0
+            .to_string()
+    } else {
+        return Err(KdcError::InvalidCnameType(cname.name_type.0.0.clone()));
+    };
+
+    kdc_config
+        .users
+        .iter()
+        .find(|user| user.username.eq_ignore_ascii_case(&username))
+        .ok_or(KdcError::ClientPrincipalUnknown(format!(
+            "the requested client principal name ({username}) is not found in KDC database",
+        )))
+}
+
+/// Validates incoming `from` and `till` values of the [KdcReqBody].
+/// Returns `auth_time` and `end_time` for the issued ticket.
+///
+/// RFC: [Generation of KRB_AS_REP Message](https://www.rfc-editor.org/rfc/rfc4120#section-3.1.3).
+fn validate_request_from_and_till(
+    from: Option<&GeneralizedTimeAsn1>,
+    till: &GeneralizedTimeAsn1,
+    max_time_skew: u64,
+) -> Result<(OffsetDateTime, OffsetDateTime), KdcError> {
+    let now = OffsetDateTime::now_utc();
+    let max_time_skew = Duration::from_secs(max_time_skew);
+
+    let auth_time = if let Some(from) = from {
+        let from = OffsetDateTime::try_from(from.0.clone())
+            .map_err(|err| KdcError::NeverValid(format!("KdcReq::from time is not valid: {err}")))?;
+        // RFC (https://www.rfc-editor.org/rfc/rfc4120#section-3.1.3):
+        // > If the requested starttime is absent, indicates a time in the past,
```

Comment:
```suggestion
        // > If the requested start time is absent, indicates a time in the past,
```

Reply:
I copied this sentence directly from the RFC. Should I change it, though?

Reply:
I think it will end up annoying if we use typos-cli at some point in the future 🤔